### ISSUE TITLE
Reproducing hash collisions

### ISSUE THREAD

**Issue by Refefer** [JohnLangford/vowpal_wabbit#1385]
Hi there,

We're trying to figure out how to reproduce hash collisions outside of Vowpal Wabbit.  In the simplest case, why would:

> 1 | a:1 a:1 b:1 
> -1 | b:1 b:1 c:1 
> 1 | a:1 a:1 c:1 
> -1 | b:1 b:1 a:1

and

> 1 | a:2 b:1 
> -1 | b:2 c:1 
> 1 | a:2 c:1 
> -1 | b:2 a:1 

end up with different losses?  

I've had a very hard time interpreting the actual code around how feature collisions are handled and the above produces two different results.  Can someone give some advice on how collisions are actually computed?

**Comment by Refefer** [comment]
I'd like to follow this up with asking for references to --adaptive, --invariant, and --normalized.  Our problem is we're trying to replicate VW performance, but we normalize and duplicate our features externally.

**Comment by arielf** [comment]
Hi @Refefer ,

Not exactly clear to me what you're trying to achieve. Trying to answer each question separately.

- In the above 2 data-sets there are no collisions.  You may use `-a` to see the hash-values each feature name gets hashed to. I get: `a` hashed to 92594, `b` hashed to 163331, `c` hashed to 185951 and `Constant` hashed to 116060.
- I think the models are a bit different for the two data-sets because (simplistic) feature counts are different so the exact weight math ends up being a bit different. `vw` is optimized for speed so it doesn't try to match feature names on the same line when counting features.
- If you normalize for scale outside `vw`, you may try `--adaptive --invariant` to disable normalization (by default all 3 enhanced-SGD options are set to on)
- If you want to force hash collisions, the most trivial way is to use a data-set with more features than 2^`--bits` (by default, `--bits` is set to 18) by increasing distinct feature count and/or reducing `--bits`.
- If you want to avoid collisions altogether, disable hashing by using numeric feature names
- As for SGD options: there's the short description provided by `--help` under `Gradient Descent options`; the pretty complex implementation in the source [gd.cc](https://github.com/JohnLangford/vowpal_wabbit/blob/master/vowpalwabbit/gd.cc) and a few papers covering the algorithms. I also found this [slide-deck which may be of help](https://fr.slideshare.net/jxieeducation/online-learning-tricks-with-vowpal-wabbit)

**Comment by Refefer** [comment]
Hi @arielf,

I appreciate the responses.  Let me try to clarify:

* In the two datasets, I'm attempting to mimic what happens when to features collide: my understanding is we 'sum' the values.  This is important for us because we're trying to move the feature quantization outside of the VW code base.  What we're trying to reverse engineer is how

> 1 | red red dress

ultimately gets converted into a sparse vector.  Do the two features, 'red', with implicit values of 1, get summed to 2 prior to the gradient descent?  How does normalization, adaptive, and invariant impact it?

> vw is optimized for speed so it doesn't try to match feature names on the same line when counting features.

Can you dig in more on that?  Are you saying that we don't actually sum features with similar names together prior to computing the derivative?  in the case of  `red red dress`, would the derivative be the sum of three features or two?  I know the count reported claims 3, but what are we actually doing inside gd.cc?

Bullets 3-6 illustrated precisely what we're doing: we perform our own hashing trick externally, map them to indexes (e.g. `3179:1`) and write them out in VW format.  The problem is in understanding how the other sgd adaptations factor in.

Thanks in advance and appreciate the response!

**Comment by MLWave** [comment]
For `--normalized` see: https://arxiv.org/abs/1305.6646 (it is done in an online manner, so you don't need an expensive pass through the full dataset to get min and max values).

**Comment by Refefer** [comment]
@MLWave Great, will read.

@arielf that's the crux.  Gradients starting at zero makes sense on first update: if weights are zero, you have symmetry in the gradient.  I'm fine with the feature counts special casing duplicates; I care more about what the actual update rule is.  When comparing `red:1 red:1 dress:1` and `red:2 dress:1`, we end up with a different gradient update when using the SGD flags.  When using vanilla SGD, we do ok: both examples are the same.
```
$ printf "1 | red:1 red:1 dress:1\n-1 | red:1 dress:1 dress:1" | vw -a --sgd --quiet
0
	red:9981:1:0	red:9981:1:0	dress:114593:1:0	Constant:116060:1:0
1
	red:9981:1:28.2843	dress:114593:1:14.1421	dress:114593:1:14.1421	Constant:116060:1:14.1421

$ printf "1 | red:2 dress:1\n-1 | red:1 dress:2" | vw -a --sgd --quiet
0
	red:9981:2:0	dress:114593:1:0	Constant:116060:1:0
1
	red:9981:1:28.2843	dress:114593:2:14.1421	Constant:116060:1:14.1421
```

Somehow, however, the gradients when using all the flags do something different in non-obvious ways:

```
$ printf "1 | red:1 red:1 dress:1\n-1 | red:1 dress:1 dress:1" | vw -a --quiet 
0
	red:9981:1:0@0	red:9981:1:0@0	dress:114593:1:0@0	Constant:116060:1:0@0
0.719302
	red:9981:1:0.230442@8	dress:114593:1:0.162953@4	dress:114593:1:0.162953@4	Constant:116060:1:0.162953@4

$ printf "1 | red:2 dress:1\n-1 | red:1 dress:2" | vw -a --quiet
	red:9981:2:0@0	dress:114593:1:0@0	Constant:116060:1:0@0
0.675840
	dress:114593:2:0.193097@4	Constant:116060:1:0.193097@4	red:9981:1:0.0965485@16
```

As an aside, I noticed the format of the output changes the `@` to a `:` when using vanilla SGD.

**Comment by arielf** [comment]
Could it be you're confusing the current (online) Weights with Sum-of-Gradients ?

When using `--sgd` the `@SumOfGradients` suffix in `--audit/-a` output gets dropped.

The general format is 4 fields separated by `:`
```
     FeatureName:HashValue:FeatureValue:CurrentWeight
```
Optionally followed (only when enhanced SGD is in effect) by per feature @SumOfGradients

**Comment by Refefer** [comment]
@arielf great observation - I was confusing the two.

That said, the resulting models _are_ different:

``
`+ printf '1 | red:1 red:1 dress:1\n-1 | red:1 dress:1 dress:1'
+ vw --noconstant

Num weight bits = 18
learning rate = 0.5
initial_t = 0
power_t = 0.5
using no cache
Reading datafile = 
num sources = 1
average  since         example        example  current  current  current
loss     last          counter         weight    label  predict features
1.000000 1.000000            1            1.0   1.0000   0.0000        3
1.917582 2.835163            2            2.0  -1.0000   0.6838        3

finished run
number of examples per pass = 2
passes used = 1
weighted example sum = 2.000000
weighted label sum = 0.000000
average loss = 1.917582
best constant = 0.000000
best constant's loss = 1.000000
total feature number = 6

+ printf '1 | red:2 dress:1\n-1 | red:1 dress:2'
+ vw --noconstant
Num weight bits = 18
learning rate = 0.5
initial_t = 0
power_t = 0.5
using no cache
Reading datafile = 
num sources = 1
average  since         example        example  current  current  current
loss     last          counter         weight    label  predict features
1.000000 1.000000            1            1.0   1.0000   0.0000        2
1.834255 2.668511            2            2.0  -1.0000   0.6336        2

finished run
number of examples per pass = 2
passes used = 1
weighted example sum = 2.000000
weighted label sum = 0.000000
average loss = 1.834255
best constant = 0.000000
best constant's loss = 1.000000
total feature number = 4
```

**Comment by MLWave** [comment]
They are different, because the number of updates is different (3 updates vs. 2 updates per example). This results in slightly different maths from the update rules.

> I think the models are a bit different for the two data-sets because (simplistic) feature counts are different so the exact weight math ends up being a bit different. vw is optimized for speed so it doesn't try to match feature names on the same line when counting features.

`--adaptive`: http://jmlr.org/papers/v12/duchi11a.html (?)
`--invariant`: https://arxiv.org/abs/1011.1576

**Comment by JohnLangford** [comment]
As @MLWave says, the update is slightly different although the predictions would be the same given a fixed model.  VW does not attempt to coalesce features with the same index by default (coalescing is a relatively expensive operation).
